composer/datasets/streaming/dataset.py

```diff
@@ -0,0 +1,379 @@
+import math
+import os
+from io import BytesIO
+from threading import Lock, Thread
+from time import sleep
+from typing import Any, Callable, Dict, Iterator, List, Optional, Sequence, Tuple
+
+import numpy as np
+from PIL import Image
+from torch import Tensor
+from torch.utils.data import IterableDataset
+from torchvision import transforms
+
+from composer.datasets.streaming.download import safe_download
+from composer.datasets.streaming.format import (StreamingDatasetIndex, bytes_to_sample_dict, get_index_basename,
+                                                get_shard_basename)
+from composer.datasets.streaming.world import World
+from composer.utils import dist
+
+
+class StreamingDataset(IterableDataset):
+    """Streaming dataset."""
+
+    def __init__(self,
+                 remote: str,
+                 local: str,
+                 shuffle: bool,
+                 decoders: Dict[str, Callable],
+                 batch_size: Optional[int] = None) -> None:
+        """Initialize with the given remote path and local cache.
+
+        Loads all the samples that are available in local cache, then starts a
+        background thread to download the rest during training. As samples are
+        added, shuffled sample selection becomes more random.
+
+        Args:
+            remote (str): Download shards from this remote directory.
+            local (str): Download shards to this local filesystem directory for reuse.
+            shuffle (bool): Whether to shuffle the samples.
+            decoders (Dict[str, Callable]): Raw bytes decoder per sample field.
+            batch_size (Optional[int]): Hint the batch size that will be used on each device's DataLoader.
+                                        Worker indices will be constructed so that there is at most 1 incomplete batch at the end of each epoch.
+        """
+        self.remote = remote
+        self.local = local
+        self.shuffle = shuffle
+        self.decoders = decoders
+        self.batch_size = batch_size
+
+        # Load the index file containing the shard metadata, either over the
+        # network or cached locally.
+        # Precomputes the shard and offset in bytes of each sample (for direct
+        # access).
+        index_filename = self._download_if_missing(get_index_basename())
+        with open(index_filename, 'rb') as fp:
+            self.index = StreamingDatasetIndex.load(fp)
+
+        # Fields, protected by the lock, relating to loading shards in the background.
+        self._lock = None
+        self._next_epoch = 0
+        self._epoch_to_todo_ids = {}
+        self._downloaded_ids = []
+        self._are_all_shards_downloaded = False
```

Comment:
Do we need to track this? To simplify the codepaths, can we assume that we are never finished downloading? (In reality, once we have all samples downloaded, then the functionality should be identical?)

Reply:
Tracking but not addressing in this PR.